[HigherOrderOp] trace subgraph with current tracer instead of creating new ones

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122972

## Motivation
Previously, exportable higher order ops starts a new make_fx session to trace the subgraphs. But this approach has many limitations:

There are many `global states`, `configs` and `mutable states` tracked by make_fx tracer, and they must be properly handled in the recursive make_fx calls. We've encountered a bunch of bugs due to this reason: 1. inductor uses a different decomposition table, but inner make_fx didn't. 2. top-level make_fx uses pre-dispatch tracing but inner make_fx didn't. 3.  metadata such 'nn_module_stack', that's tracked/mutated as states in _ModuleStackTracer are not captured for inner graph.

## Design
This PR chooses a more aggressive design: we'll use the **same tracer** to trace the subgraph. For tracer states, unless they're explicitly reset and restored in a `_reset_tracer_states_temporarily` function, they'll be preserved after tracing the subgraphs. The rationale are that 1. people generally expect the subgraphs to be treated the same as top-level graph for their new feature (e.g. in cases listed above). So by default newly added states/modified configurations should probably be preserved/tracked when tracing the subgraph. 2. There are a ton of states in the tracer but most of them won't affect the correctness of traced graph even if we allow mutations to them when tracing subgraph.

The current states listed in `_reset_tracer_states_temporarily` may NOT be complete. But the list is can be easily extended when needed.

Test Plan:
existing tests. The placeholder names for subgraph are now the variable names in parent graph when they're tracked by parent.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

ydwu4 committed:
[HigherOrderOp] trace subgraph with current tracer instead of creating new ones

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] trace subgraph with current tracer instead of creating new ones"




[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] trace subgraph with current tracer instead of creating new ones"


## Motivation
Previously, exportable higher order ops starts a new make_fx session to trace the subgraphs. But this approach has many limitations:

There are many `global states`, `configs` or `side-effects` tracked at make_fx tracer, and they must be properly handled in the recursive make_fx calls. We've encountered a bunch of bugs due to failing to do so: 1. inductor uses a different decomposition table, but inner make_fx didn't. 2. top-level make_fx uses pre-dispatch tracing but inner make_fx didn't. 3.  metadata such 'nn_module_stack', 'torch_fn' that're tracked as a side-effects in tracer are not captured for inner graph.

## Design
This PR chooses a more aggressive design: we'll use the **same tracer** to trace the subgraph. For tracer states, unless they're explicitly reset and restored in a `_reset_tracer_states_temporarily` function, they'll be preserved across different tracing sessions. The rationale are that 1. people seems to expect the subgraphs to be treated the same as top-level graph. This can be observed from the above listed cases. So by default, their newly added states/modified configurations will be preserved/tracked when tracing the subgraph. 2. There are a ton of states in the tracer but most of them won't affect the correctness of traced graph.

Note: that the current states reset in `_reset_tracer_states_temporarily` may NOT complete. We just make it easy to extend when we found some states need to be reset before tracing the subgraph.

Test Plan:
existing tests. The placeholder names for subgraph are now the variable names in parent graph when they're tracked by parent.

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] trace subgraph with current tracer instead of creating new ones"


## Motivation
Previously, exportable higher order ops starts a new make_fx session to trace the subgraphs. But this approach has many limitations:

There are many `global states`, `configs` or `side-effects` tracked by make_fx tracer, and they must be properly handled in the recursive make_fx calls. We've encountered a bunch of bugs due to this reason: 1. inductor uses a different decomposition table, but inner make_fx didn't. 2. top-level make_fx uses pre-dispatch tracing but inner make_fx didn't. 3.  metadata such 'nn_module_stack',that's tracked/mutated as states in tracer are not captured for inner graph.

## Design
This PR chooses a more aggressive design: we'll use the **same tracer** to trace the subgraph. For tracer states, unless they're explicitly reset and restored in a `_reset_tracer_states_temporarily` function, they'll be preserved after tracing the subgraphs. The rationale are that 1. people generally expect the subgraphs to be treated the same as top-level graph for their new feature (e.g. in cases listed above). So by default newly added states/modified configurations should probably be preserved/tracked when tracing the subgraph. 2. There are a ton of states in the tracer but most of them won't affect the correctness of traced graph even if we allow mutations to them when tracing subgraph.

The current states listed in `_reset_tracer_states_temporarily` may NOT be complete. But the list is can be easily extended when needed.

Test Plan:
existing tests. The placeholder names for subgraph are now the variable names in parent graph when they're tracked by parent.

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] trace subgraph with current tracer instead of creating new ones"


## Motivation
Previously, exportable higher order ops starts a new make_fx session to trace the subgraphs. But this approach has many limitations:

There are many `global states`, `configs` or `side-effects` tracked by make_fx tracer, and they must be properly handled in the recursive make_fx calls. We've encountered a bunch of bugs due to this reason: 1. inductor uses a different decomposition table, but inner make_fx didn't. 2. top-level make_fx uses pre-dispatch tracing but inner make_fx didn't. 3.  metadata such 'nn_module_stack',that's tracked/mutated as states in tracer are not captured for inner graph.

## Design
This PR chooses a more aggressive design: we'll use the **same tracer** to trace the subgraph. For tracer states, unless they're explicitly reset and restored in a `_reset_tracer_states_temporarily` function, they'll be preserved after tracing the subgraphs. The rationale are that 1. people generally expect the subgraphs to be treated the same as top-level graph for their new feature (e.g. in cases listed above). So by default newly added states/modified configurations should probably be preserved/tracked when tracing the subgraph. 2. There are a ton of states in the tracer but most of them won't affect the correctness of traced graph even if we allow mutations to them when tracing subgraph.

The current states listed in `_reset_tracer_states_temporarily` may NOT be complete. But the list is can be easily extended when needed.

Test Plan:
existing tests. The placeholder names for subgraph are now the variable names in parent graph when they're tracked by parent.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

Chillee commented:
What happens if the inner tracer interacts with a tensors from outside the HOP region?

ydwu4 commented:
> What happens if the inner tracer interacts with a tensors from outside the HOP region?

That's a good question. There are two cases:

1. If a tensor from outside the HOP region is passed as an input to the HOP region: the subgraph will just create a placeholder named _constant_input{i} when the tensor is not tracked by outer tracing context otherwise the tensor will be a normal tensor input with proxy. We temporally reset the tracer's tensor tracker and create a new tensor tracker to track the placeholder and the tensor. The subgraph will be traced in a way that's similar to inline the subgraph call but the call functioin nodes are appended to an empty graph.

2. If that outside tensor is accessed as some attributes in the HOP region: since the subgraph will create a root module, the tensor will be setattr-ed for the new root and a getattr node is created. If that tensor is accessed in another HOP region, there'll be a new root associated with that so the tensor will be setattr-ed to the new root again and create a new getattr node.

ydwu4 committed:
Update on "[HigherOrderOp] trace subgraph with current tracer instead of creating new ones"


## Motivation
Previously, exportable higher order ops starts a new make_fx session to trace the subgraphs. But this approach has many limitations:

There are many `global states`, `configs` and `mutable states` tracked by make_fx tracer, and they must be properly handled in the recursive make_fx calls. We've encountered a bunch of bugs due to this reason: 1. inductor uses a different decomposition table, but inner make_fx didn't. 2. top-level make_fx uses pre-dispatch tracing but inner make_fx didn't. 3.  metadata such 'nn_module_stack', that's tracked/mutated as states in _ModuleStackTracer are not captured for inner graph.

## Design
This PR chooses a more aggressive design: we'll use the **same tracer** to trace the subgraph. For tracer states, unless they're explicitly reset and restored in a `_reset_tracer_states_temporarily` function, they'll be preserved after tracing the subgraphs. The rationale are that 1. people generally expect the subgraphs to be treated the same as top-level graph for their new feature (e.g. in cases listed above). So by default newly added states/modified configurations should probably be preserved/tracked when tracing the subgraph. 2. There are a ton of states in the tracer but most of them won't affect the correctness of traced graph even if we allow mutations to them when tracing subgraph.

The current states listed in `_reset_tracer_states_temporarily` may NOT be complete. But the list is can be easily extended when needed.

Test Plan:
existing tests. The placeholder names for subgraph are now the variable names in parent graph when they're tracked by parent.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]
